File: solutions/1570. Dot Product of Two Sparse Vectors.cpp
class SparseVector {
public:
    vector<pair<int,int>> record; //{index, value}
    SparseVector(vector<int> &nums) {
        for (int i=0; i<nums.size(); ++i) {
            if (nums[i]) {
                record.push_back(std::make_pair(i, nums[i]));
            }
        }
    }
    
    // Return the dotProduct of two sparse vectors
    int dotProduct(SparseVector& vec) {
        int res=0, i=0, j=0;
        while (i < record.size() && j < vec.record.size()) {
            if (record[i].first == vec.record[j].first) {
                res += record[i].second * vec.record[j].second;
                i++;
                j++;
            } else if (record[i].first < vec.record[j].first) {
                i++;
            } else {
                j++;
            }
        }
        return res;
    }
};
​
// Your SparseVector object will be instantiated and called as such:
// SparseVector v1(nums1);
// SparseVector v2(nums2);
// int ans = v1.dotProduct(v2);

File: solutions/1570. Dot Product of Two Sparse Vectors.py
class SparseVector:
    def __init__(self, nums: List[int]):
        self.lst = [(idx, num) for idx, num in enumerate(nums) if num > 0]
        
    # Return the dotProduct of two sparse vectors
    # time O(L1+L2), space O(L1+L2), w/ L1, L2 = # nonzero elements in vec1, vec2
    def dotProduct(self, vec: 'SparseVector') -> int:
        i, j = 0, 0
        n, m = len(self.lst), len(vec.lst)
        ans = 0
        while i < n and j < m:
            if self.lst[i][0] == vec.lst[j][0]:
                ans += self.lst[i][1] * vec.lst[j][1]
                i += 1
                j += 1
            elif self.lst[i][0] < vec.lst[j][0]:
                i += 1
            else:
                j += 1
        return ans
        
​
# Your SparseVector object will be instantiated and called as such:
# v1 = SparseVector(nums1)
# v2 = SparseVector(nums2)
# ans = v1.dotProduct(v2)
